feat(@vercel/edge-config) Enforce an IO boundary around client APIs

gnoff opened the pull request:
Edge config is an IO library but it heavily caches the underlying IO to make reads incredibly fast. Next.js has a heuristic for prerenders that excludes IO. While we could argue that the kind of data edge-config provides is prerenderable because it has a long lifetime it is not clear that users expect this. To keep edge-config reads consistently treated as IO I've added a wrapper for the API methods which guarantees that any read will resolve in a future timeout task.

The choice of timeout is to align with React's IO semantics for what is and is not considered IO for purposes of Suspending. On edge runtimes this is unthrottled and should not add meaningful overhead. In Node.js this is at a minimum a 1ms delay if no other work takes time however since the edge config reads themselves may take longer than 1 ms often it won't be observed. Additionally as soon as one read is queued all subsequent reads that are initiated before the timeout executes will share the same boundary.

There is of course the possibility that previously scheduled events like immediates and OS IO can delay the unblocking of these APIs but if the workload is already CPU bound then reordering the resolution is not likely to meaningfully affect overall performance.

### Best reviewed with [whitespace ignored](https://github.com/vercel/storage/pull/878/files?w=1)

gnoff committed:
[Edge Config] Enforce an IO boundary around client APIs

Edge config is an IO library but it heavily caches the underlying IO to make reads incredibly fast. Next.js has a heuristic for prerenders that excludes IO. While we could argue that the kind of data edge-config provides is prerenderable because it has a long lifetime it is not clear that users expect this. To keep edge-config reads consistently treated as IO I've added a wrapper for the API methods which guarantees that any read will resolve in a future timeout task.

The choice of timeout is to align with React's IO semantics for what is and is not considered IO for purposes of Suspending. On edge runtimes this is unthrottled and should not add meaningful overhead. In Node.js this is at a minimum a 1ms delay if no other work takes time however since the edge config reads themselves may take longer than 1 ms often it won't be observed. Additionally as soon as one read is queued all subsequent reads that are initiated before the timeout executes will share the same boundary.

There is of course the possibility that previously scheduled events like immediates and OS IO can delay the unblocking of these APIs but if the workload is already CPU bound then reordering the resolution is not likely to meaningfully affect overall performance.

dferber90 commented on packages/edge-config/src/utils/io-boundary.ts:
```diff
@@ -0,0 +1,50 @@
+    // are tracked as IO by React in Server Component environments.
+    // TODO implement a Next.js specific version so we can skip this boundary checking
+    // except where needed.
+    setTimeout(() => {
```
> however since the edge config reads themselves may take longer than 1 ms often it won't be observed

We worked very hard to make latency <1ms in the majority of Edge Config reads, as we don't go over the network. p75 is definitely under 1ms.

With this change we would **_always_** be over 1ms in observed latency if any user measures `await get("foo")`. That's quite the loss.

Is there any other way to signal an IO operation to Next.js that does not require artificially delaying the execution of the operation?

And if we absolutely must merge this I would prefer skipping the setTimeout for non-Next.js to avoid punishing other frameworks for what smells a bit like Next.js leaking its implementation details to this SDK.

Can we find another way of achieving this that does not rely on the heuristic?

dferber90 commented on packages/edge-config/src/utils/io-boundary.ts:
```diff
@@ -0,0 +1,50 @@
+let blockedBoundary: null | Promise<void> = null;
+
```
_This comment is in random place so we can have a threaded discussion_

> While we could argue that the kind of data edge-config provides is prerenderable because it has a long lifetime it is not clear that users expect this.

I'm fully on board that the default expectation is for this to be IO and not part of the prerender. For, [we have this explicit `no-store` configuration option](https://github.com/vercel/storage/blob/main/packages/edge-config/src/index.ts#L278), that is used by some users of this library to treat Edge Config more like a CMS. Would that still work with the proposed changes?

dferber90 reviewed (changes requested):
Would love to explore alternatives, if there are any, as adding 1ms would double our p75 latency <picture data-single-emoji=":dramatic-chipmunk:" title=":dramatic-chipmunk:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..29MxBJMDYgzsKVwM.FatgMfWR5PO0TjwaeOMdcxLOeZpvZhI-HoQoIOA-yVXn6CxY00TujS8IqxVmo19ui8Y0HTQIt7sJufYrxr7aUPaXL9R39BlqUYCNkiVUMHV1ws0JowoMoPT8OG0xRPB7UFaK.PzMCI1EBPvCT-gX8PirzPA" alt=":dramatic-chipmunk:" align="absmiddle"></picture>

dferber90 reviewed (approved):
Approving for the snapshot release as discussed. This is still missing a changeset to be release-able as a snapshot

gnoff committed:
The boundary await has to happen before we read from any in memory cache

In Vercel the local edge config is read from filesystem but it is cached. We need the boundary to be before this because the reading from the filesystem is where things like sync IO currently happen.

dferber90 commented:
closing as #883 got merged in favor of this
